app/controllers/api/v1/organizations_controller.rb

```diff
@@ -2,8 +2,91 @@ module Api
   module V1
     class OrganizationsController < ApiController
       include Api::OrganizationsController
-
       before_action :find_organization, only: %i[users listings articles]
+      before_action :authenticate!, only: %i[create update destroy]
+      before_action :authorize_admin, only: %i[create update]
+      before_action :authorize_super_admin, only: %i[destroy]
+      after_action :verify_authorized, only: %i[create update destroy]
+
+      INDEX_ATTRIBUTES_FOR_SERIALIZATION = %i[
+        id name profile_image slug summary tag_line url
+      ].freeze
+
+      SHOW_ATTRIBUTES_FOR_SERIALIZATION = %i[
+        id username name summary twitter_username github_username url
+        location created_at profile_image slug tech_stack tag_line story
+      ].freeze
+
+      def index
+        per_page = (params[:per_page] || 30).to_i
+        num = [per_page, per_page_max].min
+        page = params[:page] || 1
+
+        organizations = Organization.select(INDEX_ATTRIBUTES_FOR_SERIALIZATION).page(page).per(num)
+
+        render json: organizations
+      end
+
+      def show
+        # This is less RESTful than having a show action that would always assume an id
+        # unless given e.g. a query parameter specifying a different lookup key.
+        # The by-username lookup is the current default behavior, this approach
+        # keeps it intact. Conventionally, the lookup would default to id-based if we keep both.
+        finder = Organization.select(SHOW_ATTRIBUTES_FOR_SERIALIZATION)
+        @organization = finder.find_by(id: params[:id_or_slug]) || finder.find_by(slug: params[:id_or_slug])
+        raise ActiveRecord::RecordNotFound unless @organization
+
+        render :show
+      rescue ArgumentError => e
+        render json: { error: e }, status: :unprocessable_entity
+      end
+
+      def update
+        set_organization
+        @organization.assign_attributes(organization_params)
+        if @organization.save
+          render json: {
+            id: @organization.id,
+            name: @organization.name,
+            profile_image: @organization.profile_image_url,
+            slug: @organization.slug,
+            summary: @organization.summary,
+            tag_line: @organization.tag_line,
+            url: @organization.url
+          }
+        else
+          render json: { error: @organization.errors_as_sentence, status: 422 }, status: :unprocessable_entity
+        end
+      rescue ArgumentError => e
+        render json: { error: e }, status: :unprocessable_entity
+      end
+
+      def destroy
+        set_organization
+        Organizations::DeleteWorker.perform_async(@organization.id, @user.id)
```

Comment:
My apologies - I think I din't portray the full story of deleting an organization in the last review and mostly focussed on the delete by super admin. 

Deleting an organization can be used in two contexts: 

Context 1: The user of the organization can delete the organization provided that the user is an admin, the organization_memberships.count == 1, articles.count.zero? and the credits.count.zero?. We enable this policies by using `authorize organization`
Context 2: The organization can be deleted _only_ by the super admin and any records that are associated to the organization will be disassociated. In this case the organization is able to be deleted because we do not enforce the policies.

For context 1 we have something like:
```
authorize organization
...
Organizations::DeleteWorker.perform_async(organization.id, current_user.id, true)
```
 where the third parameter `deleted_by_org_admin` is set to true which allows us to do certain actions like:
```
user.touch(:organization_info_updated_at)
EdgeCache::BustUser.call(user)
 # notify user that the org was deleted
 NotifyMailer.with(name: user.name, org_name: org.name, email: user.email).organization_deleted_email.deliver_now
```

For context 2 we call the worker like this: 
```
Organizations::DeleteWorker.perform_async(organization.id, current_user.id, false)
```
where the third parameter `deleted_by_org_admin` is set to false which allows us to bypass certain actions like:
```
user.touch(:organization_info_updated_at)
EdgeCache::BustUser.call(user)
 # notify user that the org was deleted
 NotifyMailer.with(name: user.name, org_name: org.name, email: user.email).organization_deleted_email.deliver_now
```

Context 2 was mainly added to allow admins to delete spammy organizations.
If you pull the latest main, you'll see we just recently added this. 

My assumption (you should probably confirm this) is that we want to cater for both these use cases through the API. Right now for the destroy route it seems like we're possibly doing a combination of both because we allow authentication by super admin but then also authorize the Organization which may cause the feature to not work as expected.

We can chat more about this if you need more clarity.

Reply:
This is super illuminating, thank you! I'll see about confirming your assumption about both use cases, but I think you're correct because Ben was just mentioning that org admins who aren't super admins should be able to delete.